tests/mtd_raw/main.c

```diff
@@ -0,0 +1,468 @@
+/*
+ * Copyright (c) 2020 ML!PA Consulting GmbH
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     tests
+ * @{
+ *
+ * @file
+ * @brief       Application for testing MTD implementations
+ *
+ * @author      Benjamin Valentin <benjamin.valentin@ml-pa.com>
+ *
+ * @}
+ */
+
+#include <assert.h>
+#include <inttypes.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "od.h"
+#include "mtd.h"
+#include "shell.h"
+#include "board.h"
+
+#ifndef MTD_NUMOF
+#ifdef MTD_0
+#define MTD_NUMOF 1
+#else
+#define MTD_NUMOF 0
+#endif
+#endif
+
+static mtd_dev_t *_get_mtd_dev(unsigned idx)
+{
+    switch (idx) {
+#ifdef MTD_0
+    case 0: return MTD_0;
+#endif
+#ifdef MTD_1
+    case 1: return MTD_1;
+#endif
+#ifdef MTD_2
+    case 2: return MTD_2;
+#endif
+#ifdef MTD_3
+    case 3: return MTD_3;
+#endif
+    }
+
+    return NULL;
+}
+
+static mtd_dev_t *_get_dev(int argc, char **argv, bool *oob)
+{
+    if (argc < 2) {
+        printf("%s: please specify the MTD device\n", argv[0]);
+        *oob = true;
+        return NULL;
+    }
+
+    unsigned idx = atoi(argv[1]);
+
+    if (idx > MTD_NUMOF) {
+        printf("%s: invalid device: %s\n", argv[0], argv[1]);
+        *oob = true;
+        return NULL;
+    }
+
+    *oob = false;
+    return _get_mtd_dev(idx);
+}
+
+static uint64_t _get_size(mtd_dev_t *dev)
+{
+    return dev->sector_count * dev->pages_per_sector * dev->page_size;
```

Comment:
If I recall correctly, non of the three is 64 bit wide. Hence, the multiplication is done in 32 bit and only afterwards extended to 64 bit, so the 32 most significant bits are always zero, even if the correct result is > 2^32 - 1.